jujuclient/setmodels_test.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package jujuclient_test
+
+import (
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/jujuclient"
+	"github.com/juju/juju/testing"
+)
+
+type ControllerModelsSuite struct {
+	testing.FakeJujuXDGDataHomeSuite
+	store          jujuclient.ClientStore
+	controllerName string
+	controller     jujuclient.ControllerDetails
+}
+
+var _ = gc.Suite(&ControllerModelsSuite{})
+
+func (s *ControllerModelsSuite) SetUpTest(c *gc.C) {
+	s.FakeJujuXDGDataHomeSuite.SetUpTest(c)
+	s.store = jujuclient.NewFileClientStore()
+	s.controllerName = "test.controller"
+	s.controller = jujuclient.ControllerDetails{
+		ControllerUUID: "test.uuid",
+		APIEndpoints:   []string{"test.api.endpoint"},
+		CACert:         "test.ca.cert",
+		Cloud:          "aws",
+		CloudRegion:    "southeastasia",
+	}
+}
+
+func (s *ControllerModelsSuite) TearDownTest(c *gc.C) {
+	s.controller = jujuclient.ControllerDetails{}
+	s.controllerName = ""
+	s.store = nil
+	s.FakeJujuXDGDataHomeSuite.TearDownTest(c)
+}
+
+func (s *ControllerModelsSuite) TestSetModelsNoController(c *gc.C) {
+	err := s.store.SetModels(s.controllerName, nil)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *ControllerModelsSuite) TestSetModelsNoControllerModels(c *gc.C) {
+	s.assertControllerStored(c)
+	err := s.store.SetModels(s.controllerName, nil)
+	c.Assert(err, jc.ErrorIsNil)
+	_, err = s.store.AllModels(s.controllerName)
+	c.Assert(err, gc.ErrorMatches, "models for controller test.controller not found")
+}
+
+func (s *ControllerModelsSuite) TestSetModelsAddOne(c *gc.C) {
+	s.assertControllerStored(c)
+	modelDetails := s.assertModelStored(c, "admin/new-model", "test.model.uuid")
+	expected := map[string]jujuclient.ModelDetails{"admin/new-model": modelDetails}
+	s.assertStoreUpdated(c, expected)
+}
+
+func (s *ControllerModelsSuite) TestSetModelsAddMany(c *gc.C) {
+	s.assertControllerStored(c)
+	expected := map[string]jujuclient.ModelDetails{
+		"admin/new-model":     s.assertModelStored(c, "admin/new-model", "test.model.uuid"),
+		"admin/another-model": s.assertModelStored(c, "admin/another-model", "test.model.uuid.2"),
+	}
+	s.assertStoreUpdated(c, expected)
+}
+
+func (s *ControllerModelsSuite) TestControllerModelsUpdate(c *gc.C) {
+	s.assertControllerStored(c)
+	expected := map[string]jujuclient.ModelDetails{
+		"admin/new-model":     s.assertModelStored(c, "admin/new-model", "test.model.uuid"),
+		"admin/another-model": s.assertModelStored(c, "admin/another-model", "test.model.uuid.2"),
+	}
+	s.assertStoreUpdated(c, expected)
+	s.assertStoreUpdated(c, expected)
```

Comment:
This was not an un-intentional duplicate. This was an additional call to ensure there is no changes to the data. I can see, however, that it's confusing. I might remove.